pubsub/src/manager.rs

```diff
@@ -1,62 +1,166 @@
 pub type TaskExecutor = Arc<dyn future::Executor<Box<dyn Future<Item = (), Error = ()> + Send>> + Send + Sync>;
 
-/// Trait used to provide unique subscription ids.
+type ActiveSubscriptions = Arc<Mutex<HashMap<SubscriptionId, oneshot::Sender<()>>>>;
+
+/// Trait used to provide unique subscription IDs.
 pub trait IdProvider {
 	/// A unique ID used to identify a subscription.
-	type Id: Copy + Clone + Default + Eq + Hash + Into<SubscriptionId> + From<SubscriptionId>;
+	type Id: Default + Into<SubscriptionId>;
 
-	/// Returns next id for the subscription.
+	/// Returns the next ID for the subscription.
 	fn next_id(&self) -> Self::Id;
 }
 
+/// Provides a thread-safe incrementing integer which
+/// can be used as a subscription ID.
+#[derive(Debug)]
+pub struct NumericIdProvider {
+	current_id: AtomicUsize,
+}
+
+impl NumericIdProvider {
+	/// Create a new NumericIdProvider.
+	pub fn new() -> Self {
+		Default::default()
+	}
+
+	/// Create a new NumericIdProvider starting from
+	/// the given ID.
+	pub fn with_id(id: AtomicUsize) -> Self {
+		Self { current_id: id }
+	}
+}
+
+impl IdProvider for NumericIdProvider {
+	type Id = usize;
+
+	fn next_id(&self) -> Self::Id {
+		self.current_id.fetch_add(1, Ordering::AcqRel)
+	}
+}
+
+impl Default for NumericIdProvider {
+	fn default() -> Self {
+		NumericIdProvider {
+			current_id: AtomicUsize::new(1),
+		}
+	}
+}
+
+/// Used to generate random strings for use as
+/// subscription IDs.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct RandomStringIdProvider {
+	len: usize,
+}
+
+impl RandomStringIdProvider {
+	/// Create a new RandomStringIdProvider.
+	pub fn new() -> Self {
+		Default::default()
+	}
+
+	/// Create a new RandomStringIdProvider, which will generate
+	/// random id strings of the given length.
+	pub fn with_len(len: usize) -> Self {
+		Self { len }
+	}
+}
+
+impl IdProvider for RandomStringIdProvider {
+	type Id = String;
+
+	fn next_id(&self) -> Self::Id {
+		let mut rng = thread_rng();
+		let id: String = iter::repeat(())
+			.map(|()| rng.sample(Alphanumeric))
+			.take(self.len)
+			.collect();
+		id
+	}
+}
+
+impl Default for RandomStringIdProvider {
+	fn default() -> Self {
+		Self { len: 16 }
+	}
+}
+
 /// Subscriptions manager.
 ///
 /// Takes care of assigning unique subscription ids and
 /// driving the sinks into completion.
 #[derive(Clone)]
-pub struct SubscriptionManager<I: Default + IdProvider> {
-	next_id: I,
-	active_subscriptions: Arc<Mutex<HashMap<I::Id, oneshot::Sender<()>>>>,
+pub struct SubscriptionManager<I: IdProvider = RandomStringIdProvider> {
+	id_provider: I,
+	active_subscriptions: ActiveSubscriptions,
 	executor: TaskExecutor, // Make generic?
 }
 
-impl<I: Default + IdProvider> SubscriptionManager<I> {
+impl<I: IdProvider> SubscriptionManager<I> {
 	/// Creates a new SubscriptionManager.
-	pub fn new(executor: TaskExecutor) -> Self {
+	pub fn new_with_id_provider(id_provider: I, executor: TaskExecutor) -> Self {
```

Comment:
```suggestion
	pub fn with_id_provider(id_provider: I, executor: TaskExecutor) -> Self {
```
AFAIK the overall convention is to skip `new_` in `with_` constructors.